stream-chat-android-client/src/main/java/io/getstream/chat/android/client/events/caching/TypingEventCache.kt

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.client.events.caching
+
+import io.getstream.chat.android.client.events.ChatEvent
+import io.getstream.chat.android.client.events.TypingStartEvent
+import io.getstream.chat.android.client.events.TypingStopEvent
+import io.getstream.chat.android.client.events.caching.TypingEventKeyValue.Companion.toStartTypingEventKeyValue
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+/**
+ * Caches [TypingStartEvent]s and automatically "cleans" stale ones
+ * after [cleanStaleEventMs] has elapsed by sending out a [TypingStopEvent]
+ * counterpart of the original event.
+ *
+ * @param coroutineScope The coroutine scope used for timed cleaning
+ * of stale jobs.
+ * @param cleanStaleEventMs The time it takes before a [TypingStartEvent] is
+ * considered stale and should be "cleaned".
+ * @param onEventFired The lambda called after a typing event has been processed
+ * and cached if needed.
+ */
+internal class TypingEventCache(
+    private val coroutineScope: CoroutineScope = CoroutineScope(DispatcherProvider.IO),
```

Comment:
This is a good observation.

My idea was to expose this so that we could pass in a `CoroutineScope` at the instantiation site. However one is not directly available in `EventParser.kt`. I would be much happier if this scope was passed in.

I wanted to avoid unnecessary scope instantiation inside of `EphemeralStartTypingEvent` because these events could be plentiful and all of them should live either shorter lives or lives that are as long as the life of `TypingEventCache` but no longer so an outer scope would prevent unnecessary instantiation.

I've added a cancel function, but I don't think that's an answer to this situation because I'm not sure if there's a good candidate that we could use to trigger it.

Advice would be appreciated here.

Reply:
Would be nice if we could link `TypingEventCache` to the socket's lifecycle.
Meaning that whenever the socket dies we cancel `TypingEventCache`.

Reply:
Planned on doing so here: https://github.com/GetStream/stream-chat-android/issues/3931